packages/marketplace/contracts/exchange/ExchangeCore.sol

```diff
@@ -295,56 +270,62 @@ abstract contract ExchangeCore is Initializable, ContextUpgradeable, TransferExe
         uint256 takeBuyAmount = newFill.rightValue;
         uint256 makeBuyAmount = newFill.leftValue;
 
-        if (((_msgSender() != msg.sender) && !nativeMeta) || ((_msgSender() == msg.sender) && !nativeOrder)) {
+        // TODO: this force me to pass from, do we want it ?
+        if (((from != msg.sender) && !nativeMeta) || ((from == msg.sender) && !nativeOrder)) {
             require(makeMatch.assetClass != LibAsset.ETH_ASSET_CLASS, "maker cannot transfer native token");
             require(takeMatch.assetClass != LibAsset.ETH_ASSET_CLASS, "taker cannot transfer native token");
         }
         if (makeMatch.assetClass == LibAsset.ETH_ASSET_CLASS) {
             require(takeMatch.assetClass != LibAsset.ETH_ASSET_CLASS, "taker cannot transfer native token");
             require(makeBuyAmount >= totalMakeValue, "not enough eth");
             if (makeBuyAmount > totalMakeValue) {
+                // TODO: from ?
                 payable(msg.sender).transferEth(makeBuyAmount - totalMakeValue);
             }
         } else if (takeMatch.assetClass == LibAsset.ETH_ASSET_CLASS) {
             require(takeBuyAmount >= totalTakeValue, "not enough eth");
             if (takeBuyAmount > totalTakeValue) {
+                // TODO: from ?
                 payable(msg.sender).transferEth(takeBuyAmount - totalTakeValue);
             }
         }
     }
 
     /// @notice parse orders with LibOrderDataGeneric parse() to get the order data, then create a new fill with setFillEmitMatch()
+    /// @param from the message sender
     /// @param orderLeft left order
     /// @param orderRight right order
     /// @return leftOrderData generic order data from left order
     /// @return rightOrderData generic order data from right order
     /// @return newFill fill result
-    function parseOrdersSetFillEmitMatch(
+    function _parseOrdersSetFillEmitMatch(
+        address from,
         LibOrder.Order memory orderLeft,
         LibOrder.Order memory orderRight
     )
-    internal
-    returns (
-        LibOrderDataGeneric.GenericOrderData memory leftOrderData,
-        LibOrderDataGeneric.GenericOrderData memory rightOrderData,
-        LibFill.FillResult memory newFill
-    )
+        internal
+        returns (
+            LibOrderDataGeneric.GenericOrderData memory leftOrderData,
+            LibOrderDataGeneric.GenericOrderData memory rightOrderData,
+            LibFill.FillResult memory newFill
+        )
     {
         bytes32 leftOrderKeyHash = LibOrder.hashKey(orderLeft);
         bytes32 rightOrderKeyHash = LibOrder.hashKey(orderRight);
 
-        address msgSender = _msgSender();
+        // TODO: this force me to pass from, do we want it ?
```

Comment:
I think this case has been introduced when rarible added directPurchase & directAcceptBid.
I think we should enforce maker != address(0)